keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1204 @@
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+      pullPolicy: IfNotPresent
+  containers:
+  - name: my-container
+    image: busybox
+    volumeMounts:
+    - mountPath: /data
+      name: oci-volume
+```
+
+### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling in the container runtime:**
+   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
+   - Extending the existing image garbage collection to not remove an OCI volume
+     image if a pod is still referencing it. 
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.
+
+**Artifacts as Subject Referrers:**
+   - Introduce new refer to image and filter for artifact type criterion/options for to be mounted artifact(s).
+   - Certain types of OCI artifacts include a subject reference. That reference
+     identifies the artifact/image for which this artifact refers. For example a
+     signature artifact could refer to a platform index, for certifying the
+     platform images, or to an SBOM artifact that refers to a platform matched
+     image. These artifacts may or may not be located on the same
+     registry/repository. The new referrers API allows for retrieving these
+     artifacts.
+
+**Validation:**
+   - Extend validation and security checks to cover new artifact types.
+
+**Storage Optimization in the container runtime:**
+   - Develop optimized storage solutions tailored for different artifact types,
+     potentially integrating with existing storage solutions or introducing new mechanisms.
+
+#### kubelet
+
+While the container runtime will be responsible of pulling and storing the OCI
+objects in the same way as for images, the kubelet still has to manage the full
+life-cycle of them. This means that some parts of the existing kubelet code can
+be reused, for example:
+
+- The logic how to ensure that an image exists on the node:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
+- The retrieval of available secrets for a pod:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988
+
+##### Pull Policy
+
+While the `imagePullPolicy` is working on container level, the introduced
+`pullPolicy` is a pod level construct. This means that we can support the same
+values `IfNotPresent`, `Always` and `Never`, but will only pull once per pod.
```

Comment:
Is the actual OCI manifest or config digest going to be recorded somewhere in the pod status or the volume itself so it is visible (via `kubectl`)?  We have this property with regular images and it is nice when pulling by tags to see what was actually pulled when the pod started.

Reply:
I added a point about that in the "CRI" section. Runtimes have to track the link between sandboxes and mounted OCI volumes so that they can provide that information to the kubelet. This is required for the kubelet image garbage collection but can be (later on) used by clients like kubectl